fix: duplicated IDs in panel

FabienMotte opened the pull request:
### Context

[CR-2709](https://algolia.atlassian.net/browse/CR-2709)

Autocomplete is rendering duplicated values for the attribute `id` (see image below, `autocomplete-0-list` ID is used in two lists). As described in the HTML specs, the value of the `id` attribute must be unique within the HTML document.

![image](https://user-images.githubusercontent.com/662153/215740717-2629b882-9e56-49fb-afc9-70c97d0782d0.png)

### Summary

ID values currently relies on the [`generateAutocompleteId()` method](https://github.com/algolia/autocomplete/blob/17a38f71f22a050ab07269d974e5fe08e29e66d8/packages/autocomplete-shared/src/generateAutocompleteId.ts), which is unique per Autocomplete instance but not per source.

#### Autocomplete panel
This PR tries to fix duplicated ID attributes in the Autocomplete panel by using the `sourceIndex` value to get a unique ID value per instance, source and item.
The ID value is described as follows: `autocomplete-${instanceId}-${sourceId}-${'list' | 'item' | 'label'}`.

#### Autocomplete search box
For the Autocomplete search box, the ID value is described as follows: `autocomplete-${instanceId}-${'input' | 'label'}`.

### Result

Outputted ID attribute values are unique.

[**→ CodeSandbox**](https://codesandbox.io/s/cranky-aryabhata-08tnbj?file=/app.tsx)

### Autocomplete instance #⁠1

#### Search box

Label ID: `autocomplete-0-label`
Input ID: `autocomplete-0-input`

#### Panel

##### Source #⁠1

List ID: `autocomplete-0-0-list`
First list item ID: `autocomplete-0-0-item-0`

##### Source  #⁠2

List ID: `autocomplete-0-1-list`
First list item ID: `autocomplete-0-1-item-5`

### Autocomplete instance #⁠2

#### Search box

Label ID: `autocomplete-1-label`
Input ID: `autocomplete-1-input`

#### Panel

##### Source #⁠1

List ID: `autocomplete-1-0-list`
First list item ID: `autocomplete-1-0-item-0`

##### Source  #⁠2

List ID: `autocomplete-1-1-list`
First list item ID: `autocomplete-1-1-item-5`

[CR-2709]: https://algolia.atlassian.net/browse/CR-2709?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

aymeric-giraudet commented on packages/autocomplete-core/src/getPropGetters.ts:
```diff
@@ -239,19 +239,32 @@ export function getPropGetters<
+        props.id,
+        sourceIndex as number
+      )}-label`,
+      id: `${getAutocompleteId(props.id, sourceIndex as number)}-list`,
```
Maybe we could add `sourceIndex?: number` to `GetListProps`'s argument type so we wouldn't need to write `as number` (which is wrong too as it can be `undefined`)
That way users could also see that something will be done if `sourceIndex` is in the props

FabienMotte replied:
Good point, thanks for the suggestion!

FabienMotte commented on packages/autocomplete-core/src/getPropGetters.ts:
```diff
@@ -239,19 +239,29 @@ export function getPropGetters<
   };
 
-  const getLabelProps: GetLabelProps = (rest) => {
+  const getAutocompleteId = (instanceId: string, sourceId?: number) => {
+    return typeof sourceId !== 'undefined'
+      ? `${instanceId}-${sourceId}`
+      : instanceId;
+  };
```
This could be extracted as an util, also not sure about the function name

aymeric-giraudet replied:
If it's used here only I think it's fine to keep it in the same file personally :D 
Simply `getId` would work for me but let's hear other opinions !

Haroenv replied:
could we consolidate between generateAutocompleteId and this function? maybe it could take more than just one argument?

aymeric-giraudet replied:
`generateAutocompleteId` is meant to be called once per autocomplete instance, prop getters are ran at every render :/
Originally they did run `generateAutocompleteId` at each render or open/close but fixed it in this commit : https://github.com/francoischalifour/autocomplete.js/commit/ce4d496d1f074d02051c1cbe1f296d2a5d6e1c1c

FabienMotte commented:
I updated the PR description and added a CodeSandbox with 2 instances, each instance with two sections (= 2 plugins).
No IDs collide, neither in the search box nor in the panel.
